### PR TITLE
feat(into_builder): implement From trait for builder

### PR DESCRIPTION
**Pull request by jonas-x**
Implement `From<Foo> for FooBuilder` to allow creating a builder from an instance.
This is useful when you want to create a new, modified, instance from an existing one.

Issue: https://github.com/colin-kiegel/rust-derive-builder/issues/170

### CODE REVIEW

**Review by TedDriggs — CHANGES_REQUESTED** [review]
There should be an opt-in for this, otherwise it might break preexisting custom impls. An opt-in, e.g. `#[builder(impl_from)]` would address that, and would provide a place for any options around the impl (e.g. how fields are populated)

File: derive_builder_core/src/builder_field.rs [hunk 1]
```diff
@@ -76,6 +76,18 @@ impl<'a> BuilderField<'a> {
         let ident = self.field_ident;
         quote! { #ident : ::derive_builder::export::core::default::Default::default(), }
     }
+
+    /// Emits a builder field initializer that initializes the field to `Some(source.field)` or
+    /// `Default::default` in the case the field is not enabled.
+    /// Used to turn an instance of `Struct` back into a builder.
+    pub fn to_builder_initializer_tokens(&self) -> TokenStream {
+        if self.field_enabled {
```
[the remark this answers is not in the document]
Comment (jonas-x) [reply derive_builder_core/src/builder_field.rs:84]: > What is the expected behavior if someone has provided custom setters?

I was not fully aware of custom setters when making these changes, been reading up on it and made some tests to verify my assumptions:

In my opinion it should not invoke the setter when converting an struct into a builder, the value should be moved, as is, from one to the other. This is because the contract of the custom setter is not known, e.g. it can expect *n* number of arguments of different types to produce the final value.

Re-setting/changing the value via the builder requires usage of the custom setter though.

Does this make sense? Have I missed something?

---

However, regarding the highlighted code where it emits the default constructor for non-enabled fields is problematic..!

Fields with `setter(skip)` and a default value/expr will be represented as `PhantomData<T>` in the builder, which prohibits setting the value in the `impl From` (if I'm not mistaken?).
The result is that these fields will regress back their default value, which of course would be very confusing if such a field had been updated during the lifetime of the struct!

Possible ways forward:

1. Define non-settable fields as `Option<T>` instead of `PhantomData<T>` and change the `build_fn` to use the default value only if the internal builder field is `None`. I encountered a test failure with this approach though in the [derive_trait](https://github.com/colin-kiegel/rust-derive-builder/blob/f58ecc719db790b4957734b6c6c7ebda30968789/derive_builder/tests/derive_trait.rs#L12-L15) test, the field type does not derive the expected traits (it didn't need to before). This is not well documented though but I assume there's a reason for the test. 

2. Document the corner case described above.

3. Hold off and regroup.


Wdyt?

[the remark this answers is not in the document]
Comment (TedDriggs) [reply derive_builder_core/src/builder_field.rs:84]: Personally, I need to take option 3 for a few weeks - both work and my personal life have gotten really busy and I can't commit the uninterrupted time needed to work through some scenarios for each case. Having some examples in Java where this is really used would be helpful, since it would provide insight into caller expectations.

**Comment by TedDriggs** [comment]
After 7 months' intermittent consideration, I don't see a way for this to be predictable for all permutations of builder options. Skipped fields and custom setters are the two most immediate places of danger, but I suspect custom build, validation, and defaulting functions would also produce surprising behavior in as-yet-unexplored edge cases.

Happily, `derive_builder` doesn't need native support for this: It's easy to write a custom `From` impl where the compiler will help make sure all fields are accounted for in the conversion.

```rust
impl From<Foo> for FooBuilder {
    fn from(v: Foo) -> Self {
        let Foo {
            first,
            second
        } = v;

        // OPTION A: Use setters
        let mut builder = FooBuilder::default();
        builder.first(first);
        builder.second(second);

        builder

        // OPTION B: Don't use setters
        FooBuilder {
            first: Some(first),
            second: Some(second),
        }
    }
}
```

By destructuring `Foo` and avoiding the use of `..`, the compiler will now 1) error if any field is left behind in `Foo`, and 2) will warn about an unused variable if a field is extracted from `Foo` but not passed through to `FooBuilder`.

If a particular project is making heavy use of this `impl From<Foo> for FooBuilder` pattern, it could define a crate-private hygienic macro to further reduce boilerplate. For example:

```rust
macro_rules! into_builder {
    ($builder:ty : $base:ident => $($field:ident),* ) => {
        impl From<$base> for $builder {
            fn from(__base: $base) -> Self {
                let $base {
                    $($field),*
                } = __base;
                
                Self {
                    $($field: Some($field)),*
                }
            }
        }
    }
}

into_builder!(FooBuilder : Foo => first, second);

```

This has the same drift protection as the more-explicit version, so adding a field to the base struct without updating the macro invocation would cause a compile error.

---

Given the lack of a single right behavior, the low demonstrated demand, and the availability of safe and explicit workarounds, I'm closing this and #170 as "Won't Do."
